core/src/main/java/org/apache/gravitino/authorization/PermissionManager.java

```diff
@@ -373,11 +376,174 @@ User revokeRolesFromUser(String metalake, List<String> roles, String user) {
+                objectExist = true;
+                Set<Privilege> updatePrivileges = Sets.newHashSet();
+                updatePrivileges.addAll(securableObject.privileges());
+                // If privileges has granted, we won't generate the new securable object.
+                if (updatePrivileges.containsAll(privileges)) {
+                  updateSecurableObjects.add(securableObject);
+                } else {
+                  updatePrivileges.addAll(privileges);
+                  SecurableObject newSecurableObject =
+                      SecurableObjects.parse(
+                          securableObject.fullName(),
+                          securableObject.type(),
+                          Lists.newArrayList(updatePrivileges));
+
+                  updateSecurableObjects.add(newSecurableObject);
+                  AuthorizationUtils.callAuthorizationPluginForMetadataObject(
+                      metalake,
+                      object,
+                      authorizationPlugin -> {
+                        authorizationPlugin.onRoleUpdated(
+                            roleEntity,
+                            RoleChange.updateSecurableObject(
+                                role, securableObject, newSecurableObject));
+                      });
+                }
+
+              } else {
+                updateSecurableObjects.add(securableObject);
+              }
+            }
+
+            // If the role doesn't contain the securable object, the role appends a new securable
+            // object.
+            if (!objectExist) {
+              SecurableObject securableObject =
+                  SecurableObjects.parse(
+                      object.fullName(), object.type(), Lists.newArrayList(privileges));
+              AuthorizationUtils.callAuthorizationPluginForMetadataObject(
+                  metalake,
+                  object,
+                  authorizationPlugin -> {
+                    authorizationPlugin.onRoleUpdated(
+                        roleEntity, RoleChange.addSecurableObject(role, securableObject));
+                  });
+              updateSecurableObjects.add(securableObject);
+            }
+
+            AuditInfo auditInfo =
+                AuditInfo.builder()
+                    .withCreator(roleEntity.auditInfo().creator())
+                    .withCreateTime(roleEntity.auditInfo().createTime())
+                    .withLastModifier(PrincipalUtils.getCurrentPrincipal().getName())
+                    .withLastModifiedTime(Instant.now())
+                    .build();
+
+            return RoleEntity.builder()
+                .withId(roleEntity.id())
+                .withName(roleEntity.name())
+                .withNamespace(roleEntity.namespace())
+                .withProperties(roleEntity.properties())
+                .withAuditInfo(auditInfo)
+                .withSecurableObjects(updateSecurableObjects)
+                .build();
+          });
```

Comment:
The update logic here is too long, you'd better split into several helper methods.

Reply:
Done.